validator/machine.go

```diff
@@ -28,19 +33,32 @@ type MachineInterface interface {
 	ProveNextStep() []byte
 }
 
+// Has a finalizer attached to remove the resolver from the global map
+type preimageResolverMarker struct {
+	id int64
+}
+
 // Holds an arbitrator machine pointer, and manages its lifetime
 type ArbitratorMachine struct {
-	ptr    *C.struct_Machine
-	frozen bool // does not allow anything that changes machine state, not cloned with the machine
+	ptr              *C.struct_Machine
+	preimageResolver *preimageResolverMarker
```

Comment:
context could theoretically have other uses than just preimageresolver. I think naming it "context" or similar makes sense. Not sure if it deserves it's own type...

Reply:
Hmm, I added the separate type because I set a finalizer on it, but I suppose that isn't necessary. I'll change it to just an int64 and rename it to context.